packages/components/src/ui/context/context-connect.ts

```diff
@@ -16,42 +16,70 @@ import { CONNECT_STATIC_NAMESPACE } from './constants';
 import { getStyledClassNameFromKey } from './get-styled-class-name-from-key';
 import type { WordPressComponentFromProps } from '.';
 
+type AcceptsTwoArgs<
+	F extends ( ...args: any ) => any,
+	ErrorMessage = never
+> = Parameters< F >[ 'length' ] extends 2 ? {} : ErrorMessage;
+
 type ContextConnectOptions = {
-	/** Defaults to `false`. */
-	memo?: boolean;
+	forwardsRef?: boolean;
 };
 
 /**
  * Forwards ref (React.ForwardRef) and "Connects" (or registers) a component
  * within the Context system under a specified namespace.
  *
- * This is an (experimental) evolution of the initial connect() HOC.
- * The hope is that we can improve render performance by removing functional
- * component wrappers.
+ * @param  Component The component to register into the Context system.
+ * @param  namespace The namespace to register the component under.
+ * @return The connected WordPressComponent
+ */
+export function contextConnect<
+	C extends ( props: any, ref: ForwardedRef< any > ) => JSX.Element | null
+>(
+	Component: C &
+		AcceptsTwoArgs<
+			C,
+			'Warning: Your component function does not take a ref as the second argument. Did you mean to use `contextConnectWithoutRef`?'
+		>,
+	namespace: string
+) {
+	return _contextConnect( Component, namespace, { forwardsRef: true } );
+}
+
+/**
+ * "Connects" (or registers) a component within the Context system under a specified namespace.
+ * Does not forward a ref.
  *
  * @param  Component The component to register into the Context system.
  * @param  namespace The namespace to register the component under.
- * @param  options
  * @return The connected WordPressComponent
  */
-export function contextConnect< P >(
-	Component: ( props: P, ref: ForwardedRef< any > ) => JSX.Element | null,
-	namespace: string,
-	options: ContextConnectOptions = {}
-): WordPressComponentFromProps< P > {
-	const { memo: memoProp = false } = options;
+export function contextConnectWithoutRef(
+	Component: ( props: any ) => JSX.Element | null,
+	namespace: string
+) {
+	return _contextConnect( Component, namespace );
+}
 
-	let WrappedComponent = forwardRef( Component );
-	if ( memoProp ) {
-		// @ts-ignore
-		WrappedComponent = memo( WrappedComponent );
-	}
+// This is an (experimental) evolution of the initial connect() HOC.
+// The hope is that we can improve render performance by removing functional
+// component wrappers.
+function _contextConnect<
+	C extends ( props: any, ref: ForwardedRef< any > ) => JSX.Element | null
+>(
+	Component: C,
+	namespace: string,
+	options?: ContextConnectOptions
+): WordPressComponentFromProps< Parameters< C >[ 0 ] > {
+	const WrappedComponent = options?.forwardsRef
+		? forwardRef< any, Parameters< C >[ 0 ] >( Component )
+		: Component;
```

Comment:
I wonder if we could simplify the code here by:
 - using `ComponentProps< C >` (or similar) instead of the more implicit `Parameters< C >[0]`
 - assigning that type directly to `WrappedComponent` — which would remove the need to specify an explicit function return type, and could potentially help with those `@ts-expect-error` comments later in the code

Reply:
Sounded hopeful but I couldn't make it happen 🙁 The blocker being, `ComponentProps< C >` needs `C` to be a `JSXElementConstructor`, while `forwardRef()` needs `C` to be a `ForwardRefRenderFunction`, and the two are not compatible 😵

Reply:
🤯 

I tried to use some `infer` magic and I came up with this


```diff
diff --git a/packages/components/src/ui/context/context-connect.ts b/packages/components/src/ui/context/context-connect.ts
index 0214ecc686..3a17897483 100644
--- a/packages/components/src/ui/context/context-connect.ts
+++ b/packages/components/src/ui/context/context-connect.ts
@@ -64,15 +64,21 @@ export function contextConnectWithoutRef< P >(
 // This is an (experimental) evolution of the initial connect() HOC.
 // The hope is that we can improve render performance by removing functional
 // component wrappers.
+type InferredComponentProps< C > = C extends (
+	props: infer P,
+	ref: ForwardedRef< any >
+) => JSX.Element | null
+	? P
+	: never;
 function _contextConnect<
 	C extends ( props: any, ref: ForwardedRef< any > ) => JSX.Element | null
 >(
 	Component: C,
 	namespace: string,
 	options?: ContextConnectOptions
-): WordPressComponentFromProps< Parameters< C >[ 0 ] > {
+): WordPressComponentFromProps< InferredComponentProps< C > > {
 	const WrappedComponent = options?.forwardsRef
-		? forwardRef< any, Parameters< C >[ 0 ] >( Component )
+		? forwardRef< any, InferredComponentProps< C > >( Component )
 		: Component;
 
 	if ( typeof namespace === 'undefined' ) {
```

But I'm not sure if it's a better solution than the current one — plus, should we wrap these props in `ComponentPropsWithoutRef` and/or `ComponentPropsWithRef` ?

It's quite entangled!

Reply:
Yeah, maybe we can add that utility type if we start to encounter it in more places.